packages/compass-import-export/src/modules/cursor-exporter.ts

```diff
@@ -0,0 +1,81 @@
+import type { AbstractCursor } from 'mongodb';
+import { EventEmitter } from 'events';
+import * as stream from 'stream';
+import { promisify } from 'util';
+import { createCSVFormatter, createJSONFormatter } from '../utils/formatters';
+const pipeline = promisify(stream.pipeline).bind(stream);
+
+export interface CursorExporterOpts {
+  cursor: AbstractCursor;
+  type: 'csv' | 'json';
+  columns: Array<string> | boolean;
+  output: stream.Writable;
+  totalNumberOfDocuments?: number;
+}
+export class CursorExporter extends EventEmitter {
+  private _cursor: AbstractCursor;
+  private _output: stream.Writable;
+  private _formatter;
+  private _columns: Array<string> | boolean;
+  private _totalNumberOfDocuments: number;
+  private _exportedDocuments = 0;
+  private _cursorStream: stream.Readable;
+  private _cancelled = false;
+  constructor(opts: CursorExporterOpts) {
+    super();
+    this._cursor = opts.cursor;
+    this._cursorStream = opts.cursor.stream();
+    this._formatter =
+      opts.type === 'csv' ? createCSVFormatter : createJSONFormatter;
+    this._output = opts.output;
+    this._columns = opts.columns ? opts.columns : true;
+    this._totalNumberOfDocuments = opts.totalNumberOfDocuments || 0;
+  }
+
+  async start(): Promise<void> {
+    try {
+      const p = pipeline(
+        this._cursorStream,
+        this.getProgressTransformStream(),
+        this.getFormatter(),
+        this._output
+      );
+      return p;
+    } catch (err) {
+      if (this._cancelled) {
+        // do nothing
+        console.log(err);
+      } else {
+        throw err;
+      }
+    }
+  }
+  cancel(): boolean {
+    this._cursorStream.unpipe();
+    this._cursorStream.emit('end');
+    this._cancelled = true;
```

Comment:
if we set `this._cancelled = true;` after `unpipe` and `emit('end')` the catch above where we check `if (this._cancelled) {` is going to potentially race and find  `this._cancelled` to be still false if, for instance `this._cursorStream.unpipe();` would cause pipeline to throw. Though i dont see that we wait it above so i'm not sure if all of this is intentional.